### ISSUE TITLE
Can't update cockpit due to backport repo not being secured

### ISSUE THREAD

**Issue by gh0sti** [cockpit-project/cockpit#17428]
### Explain what happens

1. trying to update my cockpit from 239 to 270 by following the guide posted on adding packport repo https://cockpit-project.org/running.html#debian
2. Cockpit is included in Debian since version 10 (Buster).

To get the latest version, we recommend to enable the [backports repository](https://backports.debian.org/) (as root):
. /etc/os-release
echo "deb http://deb.debian.org/debian ${VERSION_CODENAME}-backports main" > \
    /etc/apt/sources.list.d/backports.list
apt update
Install or update the package:
apt install -t ${VERSION_CODENAME}-backports cockpit
3. Error
Hit:10 https://deb.nodesource.com/node_18.x bullseye InRelease
Err:6 http://deb.debian.org/debian bullseye-backports InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 648ACFD622F3D138 NO_PUBKEY 0E98404D386FA1D9
Reading package lists... Done
W: GPG error: http://deb.debian.org/debian bullseye-backports InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 648ACFD622F3D138 NO_PUBKEY 0E98404D386FA1D9
E: The repository 'http://deb.debian.org/debian bullseye-backports InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.

### Version of Cockpit

239

### Server operating system

Debian

### Server operating system version

Bullseye 11 32bit

### What browsers are you using?

Brave

**Comment by garrett** [comment]
> on my pi

This is likely your problem: Raspbian (and Raspberry Pi OS) are not Debian, and therefore do not properly support Debian backports.

If you're using plain Debian, then it should work without having to add keys, and the packages should be able to install.

---

For what it's worth: I've tried this myself on Raspbian a while back and hit the same issue. After adding Debian's keys for the _Debian_ backport repo to Raspbian, I could upgrade Cockpit — but it was broken after trying to use Debian's version on top of Raspbian, due to some other packages that Cockpit relies upon not having proper upgrades between Raspbian and Debian.

(I now use Debian proper on my Raspberry Pi 1 instead of Raspbian / Raspberry Pi OS. And I use Fedora IoT on my Pi 4, which has the latest version of Cockpit.)

Raspbian and Raspberry Pi OS are modified versions of Debian, mainly changed in ways for what the makers of Raspberry Pi think would make a better desktop experience for kids learning to program with the Rasperry Pi (as that is the original goal of the board). Their changes aren't really great for servers and are not fully compatible with Debian... and you're likely better off using more typical distros if you're using a Pi as a server. 

However, it should be noted that you might want to upgrade your Pi's firmware with Raspberry Pi OS first, so you'll get things like USB and network booting on devices that might not support it (if they had older firmware). But, as Raspberry Pi boards boot from SD cards, it's easy to swap things around and boot from other distros without having to reinstall.

---

If you are using Debian and not Raspbian, then you shouldn't have to install keys to use backports. (And if so, we should definitely look into this... but it is outside the scope of Cockpit and would be more of a Debian problem if the Debian backports repo does not work with Debian.)

---

**TL;DR**: Are you trying to use the Debian backports repo with Raspbian or Raspberry Pi OS?

**Comment by gh0sti** [comment]
> > on my pi
> 
> This is likely your problem: Raspbian (and Raspberry Pi OS) are not Debian, and therefore do not properly support Debian backports.
> 
> If you're using plain Debian, then it should work without having to add keys, and the packages should be able to install.
> 
> For what it's worth: I've tried this myself on Raspbian a while back and hit the same issue. After adding Debian's keys for the _Debian_ backport repo to Raspbian, I could upgrade Cockpit — but it was broken after trying to use Debian's version on top of Raspbian, due to some other packages that Cockpit relies upon not having proper upgrades between Raspbian and Debian.
> 
> (I now use Debian proper on my Raspberry Pi 1 instead of Raspbian / Raspberry Pi OS. And I use Fedora IoT on my Pi 4, which has the latest version of Cockpit.)
> 
> Raspbian and Raspberry Pi OS are modified versions of Debian, mainly changed in ways for what the makers of Raspberry Pi think would make a better desktop experience for kids learning to program with the Rasperry Pi (as that is the original goal of the board). Their changes aren't really great for servers and are not fully compatible with Debian... and you're likely better off using more typical distros if you're using a Pi as a server.
> 
> However, it should be noted that you might want to upgrade your Pi's firmware with Raspberry Pi OS first, so you'll get things like USB and network booting on devices that might not support it (if they had older firmware). But, as Raspberry Pi boards boot from SD cards, it's easy to swap things around and boot from other distros without having to reinstall.
> 
> If you are using Debian and not Raspbian, then you shouldn't have to install keys to use backports. (And if so, we should definitely look into this... but it is outside the scope of Cockpit and would be more of a Debian problem if the Debian backports repo does not work with Debian.)
> 
> **TL;DR**: Are you trying to use the Debian backports repo with Raspbian or Raspberry Pi OS?

Yea I am using raspbian with debian backports repo. Doesnt raspbian overhead use less ram than debian? I might have to look into rebuilding my pi using debian

**Comment by garrett** [comment]
> Doesnt raspbian overhead use less ram than debian?

Overall, I don't think so. 

I do know that Raspbian and Rasperry Pi OS has a stripped down desktop versus other desktops like GNOME and KDE. Perhaps you're thinking about that?

If you're concerned about RAM usage I'd suggest using zram on a Pi. (It's also good basically everywhere) It's almost like free RAM. It basically sets up a swap device in part of your RAM where currently unused data in RAM is swapped into the virtual swap device when there's "pressure" to swap.

Pi 1:
```
$ free -h
              total        used        free      shared  buff/cache   available
Mem:          176Mi        46Mi        58Mi       4.0Mi        71Mi        77Mi
Swap:         355Mi        36Mi       319Mi
```

Pi 4:
```
$ free -h
               total        used        free      shared  buff/cache   available
Mem:           3.7Gi       605Mi       303Mi        38Mi       2.8Gi       2.9Gi
Swap:          3.7Gi        16Mi       3.7Gi
```

It's definitely better than a swapfile, especially on a Pi, where that's usually an SD card. Even on a system with SSD, RAM is still tons faster.

Aside: About Linux RAM usage in general: https://www.linuxatemyram.com/ (TL;DR: "Available" is the # you're looking for. Don't let "free", "buff/cache", or even "used" distract you.) As long as you're not swapping (except in zram) RAM usage is probably fine.

**Comment by dphaas2004** [comment]
For what its worth the backports are out of date a little from what im seeing on ubuntu server.  best i can get is 2.64 at the moment.  if you want newer then that from what im seeing you have to build from source.
![image](https://user-images.githubusercontent.com/56134647/178591305-274eaf4d-4287-4418-8c40-e284210b5d1c.png)

**Comment by martinpitt** [comment]
Indeed we stopped backports to Ubuntu 20.04, as newer versions don't support it any more. We removed some backwards compatibility and stopped testing on 20.04. We do continue to backport newer versions to the current 22.04 LTS, however.

**Comment by dphaas2004** [comment]
@martinpitt Thanks for this.  I can confirm after upgrade to 22.04 and using backport cockpit is at version 272.  when you say 

> "newer versions don't support it any more"

 Is this referring to newer versions of cockpit don't support 20.04?  This probably explains why I cannot build it from source on 20.04 then.  Thanks for the Info!

**Comment by martinpitt** [comment]
> Is this referring to newer versions of cockpit don't support 20.04

Correct, yes.

**Comment by gh0sti** [comment]
I was able to fix this for now by adding backports.list to my sources.list.d and installing 273-1 cockpit over 239-1
